chainer/training/updaters/multiprocess_parallel_updater.py

```diff
@@ -0,0 +1,223 @@
+import multiprocessing
+
+import six
+
+import chainer
+from chainer.dataset import convert
+from chainer import cuda
+from chainer.training.updater import StandardUpdater, _calc_loss
+
+try:
+    from cupy.cuda import nccl
+    _available = True
+except ImportError:
+    _available = False
+
+
+class _Worker(multiprocessing.Process):
+    def __init__(self, proc_id, pipe, master):
+        super(_Worker, self).__init__()
+        self.proc_id = proc_id
+        self.pipe = pipe
+        self.converter = master.converter
+        self.model = master._master
+        self.device = master._devices[proc_id]
+        self.iterator = master._mpu_iterators[proc_id]
+        self.n_devices = len(master._devices)
+
+    def setup(self):
+        print("setup %d" % self.device)
+        _, comm_id = self.pipe.recv()
+        self.comm = nccl.NcclCommunicator(self.n_devices, comm_id,
+                                          self.proc_id)
+
+        self.model.to_gpu(self.device)
+        self.reporter = chainer.reporter.Reporter()
+        self.reporter.add_observer('main', self.model)
+        print("%d setup done" % self.device)
+
+    def run(self):
+        dev = cuda.Device(self.device)
+        dev.use()
+        self.setup()
+        pp = None
+        while True:
+            job, data = self.pipe.recv()
+            if job == 'finalize':
+                dev.synchronize()
+                break
+            if job == 'update':
+                # For reducing memory
+                self.model.cleargrads()
+
+                batch = self.converter(self.iterator.next(), self.device)
+                observation = {}
+                with self.reporter.scope(observation):
+                    loss = _calc_loss(self.model, batch)
+
+                self.model.cleargrads()
+                loss.backward()
+
+                del loss
+
+                gg = self.model.gather_grads()
+                null_stream = cuda.Stream.null
+                self.comm.reduce(gg.data.ptr, gg.data.ptr, gg.size,
+                                 nccl.NCCL_FLOAT, nccl.NCCL_SUM, 0,
+                                 null_stream.ptr)
+                del gg
+                self.model.cleargrads()
+                if pp is None:
+                    pp = self.model.gather_params()
+                self.comm.bcast(pp.data.ptr, pp.size, nccl.NCCL_FLOAT, 0,
+                                null_stream.ptr)
+                self.model.scatter_params(pp)
+                pp = None
+
+                # Sending observation via pipe is too slow.
```

Comment:
I believe that this is related the issue "the reported values (loss, etc.) in subprocesses are not transferred to a main process". If this issue can't be solved, please state this explicitly in the docstring of `MultiprocessParallelUpdater` (possibly explaining the reason, that communication via pipe is too slow), removing these comments, and mark the issue as solved.